sips/sip-026/sip-026-clarity-defi-vault.md

```diff
@@ -0,0 +1,97 @@
+In Solidity, the Ethereum smart contract language, contracts can be deployed by executing Solidity code in transactions. This allows the deployment of contracts that represent vaults holding ERC-20s separately. These distinct collateral-holding contracts can be effortlessly looked up on explorers such as Etherscan, facilitating users' oversight of DeFi protocol collateralization.
+
+However, Clarity's design differs significantly. It does not allow contract deployment during transaction execution to maintain deterministic behavior and avoid infinite recursion. Consequently, representing entities that hold SIP-010 assets requires implementing logic within a single contract address.
+
+# Specification
+The proposed solution is to create a common interface for contracts that hold SIP-010 assets. While Clarity does not permit contract deployment during transactions, different entities can be logically separated within the same contract. This is used in the Arkadiko protocol contracts to differentiate between different vaults. If the Stacks Explorer implements this common interface, users will be able to view their collateral positions distinctly, providing a user experience similar to DeFi on other blockchain platforms like Ethereum and Solana.
+
+New vault IDs are generated incrementally starting from 0. Vaults are separated logically based on this ID with a numerical value.
+
+The SIPXXX Vault trait, `sipxxx-vault-trait`, has 3 functions. These functions
+do not update state, they are view-only and they allow for a common interface:
+## Trait functions
+### asset-contract
+
+`(asset-contract ((vault-id uint)) (response principal uint))`
+
+Returns the principal of the asset being held by the vault identified by `vault-id`.
+
+Returns the token type balance `token-id` of a specific principal `who` as an
+unsigned integer wrapped in an `ok` response. It has to respond with `u0` if the
+principal does not have a balance of the specified token or if no token with
+`token-id` exists. The function should never return an `err` response and is
+recommended to be defined as read-only.
+
+### holdings
+
+`(holdings ((vault-id uint) (asset <sip-010-trait>)) (response uint uint))`
+
+Returns the total amount of the underlying asset held by the vault identified by `vault-id`.
+
+### holdings-of
+
+`(holdings-of ((vault-id uint) (asset <sip-010-trait>) (owner principal)) (response uint uint))`
+
+Returns the total amount of the underlying asset held by the vault identified by
+their vault ID for `owner`. This is used when underlying assets are divided by vault id
+and are grouped by the principal of `owner`.
+
+If the implementation does not group the vault assets by owner, return `0`.
+
+
+```clarity
+(use-trait sip-010-trait .sip-010-trait-ft-standard.sip-010-trait)
+
+(define-trait vault-trait
+  (
+    (asset-contract (uint) (response principal uint))
+
+    (holdings (uint) (response uint uint))
+
+    (holdings-of (uint principal) (response uint uint))
```

Comment:
I would suggest all of the trait functions be prefaced with `vault-` because they have to share the function namespace with the rest of the contract.